code refactor for sprint 6

fzk273 opened the pull request:
no tests yet. have some questions.
add new method for HM interface
fix IMHM class based on new interface
fix IMTM to use new functionality from history
add readme for sprint 6

Commit:
code refactor for sprint 6
add new method for HM interface
fix IMHM class based on new interface
fix IMTM to use new functionality from history
add readme for sprint 6

Ivanov-s-tmn commented on src/ru/prakticum/managers/InMemoryTaskManager.java:
```diff
@@ -9,25 +9,22 @@
     private HashMap<Integer, SubTask> subtasks;
     private HashMap<Integer, Epic> epics;
-    private HistoryManager history;
+    private InMemoryHistoryManager history;
```
Используй интерфейсный тип `HistoryManager`, чтобы не привязываться к конкретной реализации `InMemoryHistoryManager`

fzk273 replied:
поправил

Ivanov-s-tmn commented on src/ru/prakticum/managers/InMemoryTaskManager.java:
```diff
@@ -9,25 +9,22 @@
         epics = new HashMap();
-        this.history = historyManager;
-
+        history = (InMemoryHistoryManager) historyManager;
```
Используй `Managers.getDefaultHistory()` для инициализации

fzk273 replied:
поправил

Ivanov-s-tmn commented on src/ru/prakticum/managers/InMemoryHistoryManager.java:
```diff
@@ -4,29 +4,106 @@
+    }
+
+    private void removeNode(Node node) {
+        if (node != null) return;
```
Проверка некорректная, нужно выходить из метода при null

fzk273 replied:
пришлось переписать метод. теперь работает корректно

Ivanov-s-tmn commented on src/ru/prakticum/managers/InMemoryHistoryManager.java:
```diff
@@ -4,29 +4,106 @@
-        if (task == null) {
-            return;
+        if (task == null) return;
+        if (history.containsKey(task.getId())) {
```
Проверка лишняя

fzk273 replied:
поправил

Ivanov-s-tmn replied:
Я тебя обманул) 
Проверка защищает нас от `NullPointerException`, если будет передан айди несуществующей задачи. Только лучше ее организовать следующим образом:
```
if (task == null) {
    return;
} 
```
Просто выйдем из метода если null, а removeNode() не обязательно вызывать под проверкой, в этом методе есть своя проверка `if (node == null) return;`

fzk273 replied:
```
        if (history.containsKey(task.getId())) {
            removeNode(history.get(task.getId()));
        }
```
если тут его не ремувить, тогда он так и останется в истории и половина тестов начнёт падать
![image](https://github.com/user-attachments/assets/baf44ef5-3656-40f2-8bc5-e9edbf11cca1)


проверку на нулл добавил

Ivanov-s-tmn replied:
Ремув убирать и не нужно, можно делать его не внутри проверки `if (history.containsKey(task.getId()))`, а просто без нее

fzk273 replied:
поправил

Ivanov-s-tmn commented on src/ru/prakticum/managers/InMemoryHistoryManager.java:
```diff
@@ -4,29 +4,106 @@
+        if (task == null) return;
+        if (history.containsKey(task.getId())) {
+            removeNode(history.get(task.getId()));
+            history.remove(task.getId());
```
Можно перенести удаление из хешмапы в метод `removeNode()` и не будет дублироваться в `remove()`

fzk273 replied:
перенёс

Ivanov-s-tmn commented on src/ru/prakticum/managers/InMemoryHistoryManager.java:
```diff
@@ -4,29 +4,106 @@
+            this.task = task;
+        }
+
+        @Override
+        public String toString() {
+            return "Node{" +
+                    "task=" + task.getId() +
+                    ", prev=" + prev +
+                    ", next=" + next +
+                    '}';
+        }
```
Тустринги тут и у менеджера истории лишние, они особо и не нужны

fzk273 replied:
я их использовал при дебаге принтами)))) удалил

Ivanov-s-tmn commented on src/ru/prakticum/managers/InMemoryHistoryManager.java:
```diff
@@ -4,29 +4,106 @@
+
+    @Override
+    public void remove(int id) {
+        if (history.containsKey(id)) {
```
Тоже лишняя проверка

fzk273 replied:
поправил

Ivanov-s-tmn commented on src/ru/prakticum/managers/InMemoryTaskManager.java:
```diff
@@ -190,9 +196,14 @@ private void updateEpicStatus(Epic epic) {
+        return (ArrayList<Task>) history.getHistory();
+    }
+
+    <T extends Task> void addToHistory(T task) {
+        history.add(task);
     }
```
Лишний метод

fzk273 replied:
удалил

Ivanov-s-tmn commented on src/ru/prakticum/managers/InMemoryTaskManager.java:
```diff
@@ -190,9 +196,14 @@ private void updateEpicStatus(Epic epic) {
     @Override
-    public List<Task> getHistory() {
-        return history.getHistory();
+    public ArrayList<Task> getHistory() {
```
Сделай возвращаемое значение `List<Task>` и не нужно будет приводить

fzk273 replied:
поправил

Commit:
code refactor for sprint 6
fix imports in tests
fix tests location to pass github check
fix linting
fix IMHM removeNode
fix IMTM constructor and fields
fix Managers

Ivanov-s-tmn commented on src/ru/prakticum/Main.java:
```diff
@@ -55,6 +55,10 @@ public static void main(String[] args) {
         Epic epic1 = manager.getEpicById(0);
         System.out.println(epic1);
 
+        //TODO я не понимаю почему последний принт истории мне всегда выдаёт лишние элементы в мапе с историей, хотя я явно удаляю этот элемент из мапы. Я уже дебагером раз 20 прошёл и так и не понял в чём проблема. need help
+        System.out.println(manager.getHistory());
+        manager.getTaskById(1);
+        System.out.println(manager.getHistory());
```
Насколько я вижу тут ты не удаляешь его, а запрашиваешь по айди `manager.getTaskById(1);`, а это считается просмотром, должен удалиться дубликат этой задачи из истории и она снова должна добавиться в конец истории.
Или ты имеешь в виду, что выводится дубликат? Приложи вывод из консоли к комментарию, визуально, я не вижу никаких проблем в логике

fzk273 replied:
это старая тудушка, забыл её удалить, я в итоге разобрался в чём проблема и пофиксил

Ivanov-s-tmn commented on test/ru/prakticum/managers/InMemoryHistoryManagerTest.java:
```diff
@@ -0,0 +1,90 @@
+        historyManager.remove(task.getId());
+        Assertions.assertEquals(0, historyManager.getHistory().size());
+    }
+    //TODO непонимаю как протестить класс Node не могу никак импортнуть его сюда, поскольку он находится в другом классе. нид хэлп
```
А зачем он нужен? Покрой тестами удаление из разных частей истории (из начала, из середины и с конца), важно проверить порядок задач в истории после удаления из разных частей

fzk273 replied:
в итоге так и сделал. спасибо

Ivanov-s-tmn commented on java-kanban.iml:
файлы с расширением .iml лучше добавить в исключения gitignore файла, они не генерируются локально и не нужны в репозитории

fzk273 replied:
убрал из трека гита

Commit:
code refactor for sprint 6
add tests for history manager

Ivanov-s-tmn commented on test/ru/prakticum/managers/InMemoryHistoryManagerTest.java:
```diff
@@ -85,6 +82,38 @@ void checkTaskIsRemovedFromHistory() {
+        Assertions.assertEquals(3, historyManager.getHistory().size());
+        historyManager.remove(epic.getId());
+        Assertions.assertEquals(2, historyManager.getHistory().size());
+        Assertions.assertEquals(subTask, historyManager.getHistory().get(1));
```
Желательно проверить полное совпадение истории с порядком, как пример:
`Assertions.assertEquals(List.of(task, subtask), historyManager.getHistory());`
Достаточно будет одной этой проверки в этих тестах, она проверит и размер истории и порядок в истории

fzk273 replied:
поправил

Commit:
fix test removeTaskFromTheMiddleOfHistory
fix add method in history manager
